Developer guide - Add use cases

zzawook opened the pull request:
Closes #50 
Add use cases to developer guide

dwsc37 commented on docs/DeveloperGuide.md:
```diff
@@ -290,43 +290,277 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
+4. User confirms the details.
+5. LMS saves the student's address and notifies the user.
 
-### Non-Functional Requirements
```
I think the placeholders for the NFRs and glossary were removed

zzawook replied:
Recovered the placeholders with new commit!

dwsc37 commented on docs/DeveloperGuide.md:
```diff
@@ -290,43 +290,277 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
 (For all use cases below, the **System** is the `AddressBook` and the **Actor** is the `user`, unless specified otherwise)
 
-**Use case: Delete a person**
+#### Use case: Save a student's contact
```
Looks good overall, but I think everyone should look through it to ensure its ok

dwsc37 commented on docs/DeveloperGuide.md:
```diff
@@ -290,30 +290,277 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
 (For all use cases below, the **System** is the `AddressBook` and the **Actor** is the `user`, unless specified otherwise)
 
-**Use case: Delete a person**
+#### Use case: Save a student's contact
```
This should be combined with the other two use cases into just "Add a student"

dwsc37 commented on docs/DeveloperGuide.md:
```diff
@@ -290,30 +290,277 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
-4.  AddressBook deletes the person
+1. User enters command to save a student's contact.
+2. LMS parses the contact details from the user's command.
+3. LMS requests for confirmation.
+4. User confirms the details.
```
I don't think we should have confirmations for these commands

dwsc37 commented on docs/DeveloperGuide.md:
```diff
@@ -290,30 +290,277 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
+  * Use case resumes from step 3.
 
-  Use case ends.
+* \*a. At any time, User chooses to cancel the saving process.
+  * \*a1. LMS requests to confirm the cancellation of the process.
+  * \*a2. User confirms the cancellation of the process.
```
Do we need to include this? There isn't really any cancellation of the process.

dwsc37 commented on docs/DeveloperGuide.md:
```diff
@@ -290,30 +290,277 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
+
+1. User enters command to add a piano piece to a student.
+2. LMS parses the piano piece details from the user's command.
+3. LMS requests for confirmation.
+4. User confirms the addition.
```
Ditto

dwsc37 commented on docs/DeveloperGuide.md:
```diff
@@ -290,30 +290,277 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
+  * Steps 2a1-2a2 are repeated until the data entered are correct.
+  * Use case resumes from step 3.
+
+* \*a. At any time, User chooses to cancel the addition process.
+  * \*a1. LMS requests to confirm the cancellation of the process.
+  * \*a2. User confirms the cancellation of the process.
```
Ditto

dwsc37 commented on docs/DeveloperGuide.md:
```diff
@@ -290,30 +290,277 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
+
+---
+
+#### Use case: View a student's grade level
+
+**MSS**
+
+1. User enters command to view a student's grade level.
+2. LMS parses the student's details from the user's command.
+3. LMS retrieves the student's grade level from the database.
+4. LMS displays the student's grade level to the user.
+
+   Use case ends.
+
+**Extensions**
+
+* 3a. LMS detects an error while retrieving the grade level.
+  * 3a1. LMS shows an error message.
+  * Use case ends.
+
+* \*a. At any time, User chooses to cancel the process.
+  * \*a1. LMS requests to confirm the cancellation of the process.
+  * \*a2. User confirms the cancellation of the process.
```
I think this is should be changed to "View student's detailed information"

dwsc37 commented on docs/DeveloperGuide.md:
```diff
@@ -290,30 +290,277 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
+
+1. User enters command to cancel a lesson session.
+2. LMS parses the lesson session details from the user's command.
+3. LMS requests for confirmation.
+4. User confirms the cancellation.
```
Ditto

dwsc37 commented on docs/DeveloperGuide.md:
```diff
@@ -290,30 +290,277 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
+  * Steps 2a1-2a2 are repeated until the data entered are correct.
+  * Use case resumes from step 3.
+
+* \*a. At any time, User chooses to cancel the cancellation process.
+  * \*a1. LMS requests to confirm the cancellation of the process.
+  * \*a2. User confirms the cancellation of the process.
```
Ditto

dwsc37 commented on docs/DeveloperGuide.md:
```diff
@@ -290,30 +290,277 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
+
+1. User enters command to schedule a make-up lesson.
+2. LMS parses the make-up lesson details from the user's command.
+3. LMS requests for confirmation.
+4. User confirms the scheduling.
```
Ditto

dwsc37 commented on docs/DeveloperGuide.md:
```diff
@@ -290,30 +290,277 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
+  * Steps 2a1-2a2 are repeated until the data entered are correct.
+  * Use case resumes from step 3.
+
+* \*a. At any time, User chooses to cancel the scheduling process.
+  * \*a1. LMS requests to confirm the cancellation of the process.
+  * \*a2. User confirms the cancellation of the process.
```
Ditto

dwsc37 commented on docs/DeveloperGuide.md:
```diff
@@ -290,30 +290,277 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
+
+1. User enters command to delete a student.
+2. LMS parses the student details from the user's command.
+3. LMS requests for confirmation.
+4. User confirms the deletion.
```
Ditto

dwsc37 commented on docs/DeveloperGuide.md:
```diff
@@ -290,30 +290,277 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
-3.  User requests to delete a specific person in the list
-4.  AddressBook deletes the person
+1. User enters command to save a student's contact.
+2. LMS parses the contact details from the user's command.
```
All instances of LMS should be replaced with KeyContacts

dwsc37 commented on docs/DeveloperGuide.md:
```diff
@@ -290,30 +290,277 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
+2. LMS parses the student details from the user's command.
+3. LMS requests for confirmation.
+4. User confirms the deletion.
+5. LMS deletes the student from the database and notifies the user.
```
Should not refer to it as a 'database'

dwsc37 commented on docs/DeveloperGuide.md:
```diff
@@ -290,30 +290,277 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
+  * Steps 2a1-2a2 are repeated until the data entered are correct.
+  * Use case resumes from step 3.
+
+* \*a. At any time, User chooses to cancel the deletion process.
+  * \*a1. LMS requests to confirm the cancellation of the process.
+  * \*a2. User confirms the cancellation of the process.
+  * Use case ends.
```
Ditto

dwsc37 commented on docs/DeveloperGuide.md:
```diff
@@ -290,30 +290,277 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
+
+---
+
+#### Use case: Save and retrieve data after application restart
+
+**MSS**
+
+1. User enters data into the application.
+2. User chooses to save the data.
+3. LMS saves the data to persistent storage.
+4. User turns off the application.
+5. User restarts the application.
+6. LMS retrieves the data from persistent storage and displays it to the user.
+
+   Use case ends.
+
+**Extensions**
+
+* 3a. LMS detects an error while saving the data.
+  * 3a1. LMS shows an error message.
+  * Use case ends.
+
+* 6a. LMS detects an error while retrieving the data.
+  * 6a1. LMS shows an error message.
+  * Use case ends.
```
This should not be here, because saving data and loading data is done automatically.

zzawook replied:
I personally think saving and retrieving data should be in use cases. It's an interaction between user and system that user expects as granted

dwsc37 replied:
Use case should display only externally visible behaviour, and should involve interactions between user and system. Saving data is done solely by the system, there is no user interaction involve. Additionally, saving data is done behind the scenes by the system, with no response to the user.

dwsc37 reviewed (changes requested):
Left some comments for changes.

Additionally, the intention is to add use cases for the full project, not just the MVP. 
I think we should discuss as a group on the other use cases to add (grouping, schedule viewing etc.)

zzawook commented:
> Left some comments for changes.
> 
> Additionally, the intention is to add use cases for the full project, not just the MVP. I think we should discuss as a group on the other use cases to add (grouping, schedule viewing etc.)

I made corresponding changes, as suggested, so check it out!
- modified 'database' to 'storage'
- merged duplicate use cases
- Removed confirmation logic in all operations. (But wouldn't it be better to have one? idk)

dwsc37 commented:
I think somehow the bin folder got accidentally pushed?

zzawook commented:
> I think somehow the bin folder got accidentally pushed?

fixed

dwsc37 reviewed (approved):
LGTM
